File: a2zSheet/Trie/XorQueries.cpp
struct Node{
    Node* links[2];
};

class Trie{
    Node* root;
    public:
    Trie(){
        root = new Node();
    }
    void insert(int num){
        Node* temp = root;
        for(int i = 31;i>=0;i--){
            int bit = (num>>i)&1;
            if(!temp->links[bit]){
                temp->links[bit] = new Node();
            }
            temp = temp->links[bit];
        }
    }

    int getMax(int num){
        Node* temp = root;
        int maxi = 0;
        for(int i = 31;i>=0;i--){
            int bit = (num>>i)&1;
            if(temp->links[!bit]){
                maxi |= (1<<i);
                temp = temp->links[!bit];
            }
            else{
                temp = temp->links[bit];
            }
            
        }
        return maxi;
    }
};
struct ds{
    int x;
    int m;
    int ind;
};
class Solution {
public:
    vector<int> maximizeXor(vector<int>& nums, vector<vector<int>>& q) {
        sort(nums.begin(),nums.end());
        vector<ds>Q;
        for(int i=0;i<q.size();i++){
            ds d;
            d.m = q[i][1];
            d.x = q[i][0];
            d.ind = i;
            Q.push_back(d);
        }
        
        sort(Q.begin(),Q.end(),[&](ds d1,ds d2){
            return d1.m<d2.m;
        });

        vector<int>ans(q.size(),0);
        int n = nums.size();
        int ind = 0;
        
        Trie t;
        
        for(auto &c:Q){
            int x = c.x;
            int m = c.m;
            int qInd = c.ind;

            while(ind<n && nums[ind]<=m){
                t.insert(nums[ind]);
                ind++;
            }
            if(ind==0)ans[qInd] = -1;
            else ans[qInd] = t.getMax(x);
        }
        return ans;
    }
};

File: a2zSheet/Trie/XorQueries.go
package main

import (
	"sort"
)

type Node struct {
	links [2]*Node
}

type Trie struct {
	root *Node
}

func Constructor() Trie {
	return Trie{root: &Node{}}
}

func (t *Trie) Insert(num int) {
	temp := t.root
	for i := 31; i >= 0; i-- {
		bit := (num >> i) & 1
		if temp.links[bit] == nil {
			temp.links[bit] = &Node{}
		}
		temp = temp.links[bit]
	}
}

func (t *Trie) GetMax(num int) int {
	temp := t.root
	maxi := 0
	for i := 31; i >= 0; i-- {
		bit := ((num >> i) & 1)
		if temp.links[1-bit] != nil {
			maxi |= (1 << i)
			temp = temp.links[1-bit]
		} else {
			temp = temp.links[bit]
		}
	}

	return maxi
}

type ds struct {
	x   int
	m   int
	ind int
}

func maximizeXor(nums []int, queries [][]int) []int {

	sort.Ints(nums)

	Q := make([]ds, len(queries))
	for i, q := range queries {
		Q[i] = ds{q[0], q[1], i}
	}

	sort.Slice(Q, func(i, j int) bool {
		return Q[i].m < Q[j].m
	})

	// fmt.Print(Q)
	ans := make([]int, len(queries))
	n := len(nums)
	ind := 0

	t := Constructor()

	for _, c := range Q {
		x := c.x
		m := c.m
		qInd := c.ind

		for ind < n && nums[ind] <= m {
			t.Insert(nums[ind])
			ind++
		}
		if ind == 0 {
			ans[qInd] = -1
		} else {
			ans[qInd] = t.GetMax(x)
		}
	}
	return ans

}
